### PR TITLE
TN-3117 out of window proms UI changes

### PR DESCRIPTION
**Pull request by achen2401**
https://jira.movember.com/browse/TN-3117

- a new required Visit field is added to the manual entry modal
- If the entered questionnaire completion date is not within the time window of the selected visit, a warning message will appear and tell user as such.  However, the user **will not** be prevented to continue. Dates displayed to the user will be in **local** date/time (down to the minutes)
- a new `actual` (date actually completed) query string is passed on to the backend api when applicable 
- `actual` date is determined if the date comes after the visit starts, the end date of the visit minus one day is passed along; if the date comes before the visit starts, the start date of the visit is passed along
example url that will be passed on from the frontend: ` /api/present-needed?subject_id=3067&entry_method=paper&authored=2022-01-21T01:16:59Z&actual=2022-01-20T01:16:59Z`

### CODE REVIEW

**Comment by achen2401** [comment]
@pbugni & @ivan-c Will you have a minute to try out my changes?  Let me know if you see any issue?

**Comment by mcjustin** [comment]
@achen2401  I thought this was contingent on @pbugni  adding some new services?

**Comment by achen2401** [comment]
@mcjustin Yes, you are right.  This PR only addresses the frontend changes.

File: portal/static/js/src/modules/TnthAjax.js [hunk 1]
```diff
@@ -1084,6 +1084,24 @@ export default { /*global $ */
...
+        if (!userId) {
+            callback({error: i18next.t("User id is required.")});
+            return false;
+        }
+        this.sendRequest("/api/patient/" + userId + "/timeline", "GET", userId, params, function(data) {
+            if (data) {
+                if (!data.error) {
+                    callback(data);
+                } else {
+                    callback({"error": i18next.t("Error occurred retrieving assessment timeline.")});
```
Comment (ivan-c) [inline portal/static/js/src/modules/TnthAjax.js:1098]: >Error occurred retrieving assessment timeline.

If we want to translate this string, could you replace this error message with an existing (more generic) one that's already translated?

**Review by ivan-c — APPROVED** [review]
Looks good!

**Comment by pbugni** [comment]
> @mcjustin Yes, you are right. This PR only addresses the frontend changes.

@achen2401 I was under the impression from our session on Thursday, that the existing API, namely `/api/patient/<id>/timeline` delivered all you needed for this feature.  Is there something lacking?

**Comment by achen2401** [comment]
@pbugni sorry, I am merely pointing out that there are other backend changes that still have to made based on the [specs](https://promowiki.movember.com/pages/viewpage.action?spaceKey=IR&title=Capturing+%27out+of+window%27+PROMs), i.e. **5.b.iii, 5.b.iv** This PR is purely for frontend changes.

**Review by pbugni — CHANGES_REQUESTED** [review]
We do NOT want to transmit "visit" information through the AE and back to the portal, but rather just dates.
- the `authored` date must be within the valid timeframe of the requested visit.  for `actual` dates outside the valid timeframe, it should be the closer fencepost (i.e. the end of the visit if it happened after, the start of the visit if it happened before).  the `authored` date will continue to be used to look up *which* visit the response belongs to.
- the `actual` date need only be included when the user selects a date outside the valid visit window.  it will just ride along, be included in the parameters passed to AE, and the AE will package w/i the QNR.

**Review by pbugni — APPROVED** [review]
@achen2401 looks great.  Please update the PR comment to match what is now being passed (i.e. dates and not visit info).  Thanks!

**Pull request merged**